### PR TITLE
Add support for `date_nanos` and tests.

### PR DESCRIPTION
**Pull request by Yury-Fridlyand**
### Description
Restore support for `date_nanos` type. Initially was added in https://github.com/opensearch-project/sql/pull/360 and removed in https://github.com/opensearch-project/sql/pull/1314 by mistake.
 
### Issues Resolved
https://github.com/opensearch-project/sql/issues/1954
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
  - [ ] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by MitchellGale — APPROVED** [review]

File: integ-test/src/test/resources/indexDefinitions/datatypes_index_mapping.json [hunk 1]
```diff
@@ -13,10 +13,14 @@
       "binary_value": {
         "type": "binary"
       },
-      "date_value":  {
+      "date_value": {
         "type" : "date",
         "format": "yyyy-MM-dd HH:mm:ss"
       },
+      "date_nanos_value": {
+        "type" : "date_nanos",
```
Comment (GumpacG) [inline integ-test/src/test/resources/indexDefinitions/datatypes_index_mapping.json:21]: Can you add tests without format being specified in the mapping? Test default formats as per specified in https://opensearch.org/docs/latest/field-types/supported-field-types/date-nanos/#:~:text=If%20you%20don%E2%80%99t%20specify%20the%20format%20when%20mapping%20a%20field%20as%20date_nanos%2C%20the%20default%20format%20is%20strict_date_optional_time%7C%7Cepoch_millis%20that%20lets%20you%20pass%20values%20in%20either%20strict_date_optional_time%20or%20epoch_millis%20format

Comment (Yury-Fridlyand) [reply integ-test/src/test/resources/indexDefinitions/datatypes_index_mapping.json]: Fixed in 7a04f82b7, thanks. `date_nanos` with format tested in another test.

**Review by matthewryanwells — APPROVED** [review]

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java [hunk 2]
```diff
@@ -120,6 +122,31 @@ public void testNumericFormats() {
...
+  @SneakyThrows
+  public void testDateNanosWithNanos() {
+    String query = String.format("SELECT date_nanos_value"
+        + " FROM %s", TEST_INDEX_DATATYPE_NONNUMERIC);
+    JSONObject result = executeQuery(query);
+    verifySchema(result,
+        schema("date_nanos_value", null, "timestamp"));
+    verifyDataRows(result,
+        rows("2019-03-24 01:34:46.123456789"));
+  }
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:148]: We need to ensure that we can convert date_nanoes to date type.

Comment (acarbonetto) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:148]: Are there any additional requirements on date_nanos?  Should that type work with any other existing function, or does it first need to be cast to `date` type?

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java [hunk 3]
```diff
@@ -120,6 +122,31 @@ public void testNumericFormats() {
...
   }
 
+  @Test
+  @SneakyThrows
+  public void testDateNanosWithFormats() {
+    String query = String.format("SELECT hour_minute_second_OR_t_time"
+        + " FROM %s", TEST_INDEX_DATE_FORMATS);
+    JSONObject result = executeQuery(query);
+    verifySchema(result,
+        schema("hour_minute_second_OR_t_time", null, "time"));
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:132]: can you change this to `hour_minute_second_millis` instead?  The point of using the `date_nano` type is to include fractions of a second.

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java [hunk 4]
```diff
@@ -120,6 +122,31 @@ public void testNumericFormats() {
         rows("1970-01-02 03:55:00", "1970-01-01 00:01:40.5"));
   }
 
+  @Test
+  @SneakyThrows
+  public void testDateNanosWithFormats() {
```
Comment (acarbonetto) [inline integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:127]: Maybe add a test that uses custom format with fractions of a second?

**Comment by acarbonetto** [comment]
@Yury-Fridlyand we should add tests to parallel the expected features of date_nanos.  
We should add a test that uses `<` or `>` to compare a `date_nano` type with a date or time.

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java (hunk 2 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:148]: No need for a cast, it is converted to the same internal types as `date` - `OpenSearchDateType`.

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java (hunk 3 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:132]: I use `t_time` here, it includes up to 9 digits of fsp.

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java (hunk 4 above)
Comment (Yury-Fridlyand) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:127]: See next test (line 140)

**Comment by Yury-Fridlyand** [comment]
> @Yury-Fridlyand we should add tests to parallel the expected features of date_nanos.
> 
> * We should add a test that uses `<` or `>` to compare a `date_nano` type with a date or time.
> * We should include a test with `date_nano` type in the ORDER BY clause.
>   Reference: https://opensearch.org/docs/latest/field-types/supported-field-types/date-nanos/

Added tests and rebased on upstream/main.

**Review by GumpacG — APPROVED** [review]

File: integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java (hunk 3 above)
Comment (acarbonetto) [reply integ-test/src/test/java/org/opensearch/sql/sql/DateTimeFormatsIT.java:132]: but the verifyDataRows doesn't include any units of precision?  Maybe include data that has fractions of a second?

**Review by acarbonetto — APPROVED** [review]

**Pull request merged**